core/state/statedb.go

```diff
@@ -808,6 +846,10 @@ func (s *StateDB) IntermediateRoot(deleteEmptyObjects bool) common.Hash {
 			obj.updateRoot(s.db)
 			s.updateStateObject(obj)
 		}
+		usedAddresses = append(usedAddresses, addr)
+	}
```

Comment:
If we were to use two separate routines for accounts/storage respectively, and had two separate `Pause` functions, we could do it in separate phases: 
1. Pause the storage worker, 
2. Do the storage hashing
3. Pause the account worker
4. Do the account hashing

It would look something like this: 
```golang
	// Update the root of all stateobjects in pending
	for addr := range s.stateObjectsPending {
		obj := s.stateObjects[addr]
		if !obj.deleted {
			obj.updateRoot(s.db)
		}
	}
	if s.prefetcher != nil{
		s.prefetcher.PauseAccounts()
		if trie := s.prefetcher.GetTrie(s.originalRoot); trie != nil {
			s.trie = trie
		}
	}
	// Now update the account trie
	for addr := range s.stateObjectsPending {
		obj := s.stateObjects[addr]
		if obj.deleted {
			s.deleteStateObject(obj)
		} else {
			s.updateStateObject(obj)
		}
		usedAddresses = append(usedAddresses, addr)
	}
```

Reply:
Here's a working implementation, perhaps still with some rough edges: https://github.com/holiman/go-ethereum/pull/23/files

Reply:
If the storage hashing takes ~5-10ms of mostly-non-disk-accessing processing, then an additional 5-10ms given to the prefetcher for the remainder accounts might improve the account trie performance

Reply:
I'm wondering whether we should go all in and have 1 prefetcher per trie (root). That would maximize concurrency and make it more generic in that there's no account/storage distinction. Will try to have a stab at it. Maybe we should prep a new database with the snapshots generated.